### ISSUE TITLE
when using YOLO to detect video, streaming; how can I extract result of each frame when yolo is predicting instead of waiting for detect completed?

### ISSUE THREAD

**Issue by dinhthihuyen** [ultralytics/ultralytics#6046]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I use YOLO to detect video / stream source:
from ultralytics import YOLO
model = YOLO("yolov8.pt")
result = model("video.mp4")  # or result = model("rtsp://localhost:8854/video")
for detect_info in result:
     msg = detect_info.boxes.data
     nats_client.publish('message', msg)

When model is predicting, I want to get result, create a message and publish it to another nats client.
However, with above command, I can only publish message when detect is done.
I think about changing source code of YOLOv8, when pass a function as argument, call it when meet "yield from self.results" in /engine/predictor.py; or write a function to read each frame from source video (stream). However, I do not think it is optimal.
Is there any other ways that is more optimal?
Thanks so much for your support!!!

**Comment by RizwanMunawar** [comment]
@dinhthihuyen You can utilize the provided code to send a message using the `nats_client` for each frame.

```python
import cv2
import sys
from ultralytics import YOLO
from ultralytics.utils.plotting import Annotator

video_path = "path/to/video.mp4"
cap = cv2.VideoCapture(video_path)

if not cap.isOpened():
    print("Error reading video file")
    sys.exit()


while cap.isOpened():
    success, frame = cap.read()
   
    if success:
        results = model.predict(frame)
        boxes = results[0].boxes.xywh.cpu()
        clss = results[0].boxes.cls.cpu().tolist()
        names = results[0].names
        confs = results[0].boxes.conf.float().cpu().tolist()
        
        annotator = Annotator(frame, line_width=2, example=str(names))
        
        for box, cls in zip(boxes, clss):
            x, y, w, h = box
            label = str(names[cls])
            x1, y1, x2, y2 = x - w / 2, y - h / 2, x + w / 2, y + h / 2
            annotator.box_label([x1, y1, x2, y2], label=names[int(cls)],
                                color=(255, 0, 255))

        nats_client.publish('message', boxes)
        cv2.imshow("YOLOv8 Detection", frame)

        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        break

cap.release()
cv2.destroyAllWindows()
```

Hope this helps! Let me know if you have any additional questions.
Thanks

**Comment by dinhthihuyen** [comment]
Hello @RizwanMunawar, 
Thanks so much for your support!! I did it well with video, however I got troubles when do the same thing with stream source.
I need to read stream source, handle predict with each frame and return another rtsp source with output.
import subprocess as sp

rtsp_cmd = ['ffmpeg', '-y', '-f', 'rawvideo', '-s', f'{640}x{360}', '-pixel_format', 'bgr24', '-r', f'{30}', '-i', '-', '-pix_fmt', 'yuv420p', '-c:v', 'libx264', '-bufsize', '64M', '-maxrate', '4M', '-rtsp_transport', 'tcp', '-f', 'rtsp', 'rtsp://localhost:8554/output'
process = sp.Popen(self.rtsp_cmd, stdin=sp.PIPE)

stream_path = "rtsp://localhost:8554/input"
cap = cv2.VideoCapture(stream_path )

if not cap.isOpened():
    print("Error reading video file")
    sys.exit()

while cap.isOpened():
    success, frame = cap.read()
   
    if success:
        results = model.predict(frame)
        annotated_frame = results[0].plot()
        process.stdin.write(annotated_frame)

        if cv2.waitKey(1) & 0xFF == ord("q"):
            break
    else:
        break

cap.release()

However, streaming stop at frame 1, which I do not expect!
How can I handle it? Thanks so much for your support!!!
